proposals/0043-prefetch.md

```diff
@@ -0,0 +1,147 @@
+- Enable prefetching via an attribute/hint
+- Enable prefetching for all links by default (required by View transitions)
+- Disable prefetching if all links are enabled by default
+- Different prefetching strategies (click, hover, viewport, etc)
+- Only add JS if using prefetching
+
+# Non-Goals
+
+- Prefetch cache invalidation (Browser relies on cache control)
+- Prefetch external links
+
+# Detailed Design
+
+A prefetch script for client-side is required. It should only be included if the `prefetch` config is truthy. The script can be injected through the `injectScript` integration API.
+
+### Config
+
+The prefetch configuration is a top-level Astro config:
+
+```js
+// default value if `prefetch: true` (prefetch is not enabled by default)
+prefetch: {
+  // Whether to prefetch all links by default
+  prefetchAll: false,
+  // Default prefetch strategy for `data-astro-prefetch` (no value specified).
+  defaultStrategy: 'hover' // accepts 'tap' and 'viewport' too
+}
+```
+
+If View Transitions is used in Astro, the default value of the `prefetch` config (if user not configured) is `{ prefetchAll: true }`. The user can configure `false`, `{ prefetchAll: false }`, etc if they want to override this default.
+
+### Client script
+
+The script should attach listeners on initialization for different prefetch strategies:
+
+- `tap`: Calls `fetch()` on `touchstart` or `mousedown` events
+- `hover`: Calls `fetch()` on `focusin`+`focusout` or `mouseenter`+`mouseleave` events. Hover detection kicks in after 80ms delay.
+- `viewport`: Creates a `<link rel="prefetch">` on enter viewport. It has lower priority than `fetch()` to not clog up the request. Intersection detection kicks in after 300ms.
+
+Additional rules:
+
+- Prefetched links should not be fetched again (e.g. hovering on a link twice)
+- The strategy should only apply when `data-astro-prefetch`'s value matches
+- If `data-astro-prefetch` has no value, use the configured `defaultStrategy`
+- If `prefetchAll` is enabled, apply `defaultStrategy` for all links
+
+Notes:
+- `fetch()` has higher priority than `<link rel="prefetch">` when prefetching
+
+### Programmatic API
+
+The client script would have an internal `prefetch` function that we can expose to the `astro:prefetch` module:
+
+```ts
+export declare function prefetch(url: string, opts?: { with?: 'link' | 'fetch' }): void
+```
+
+This module can only be imported if the `prefetch` config is enabled.
+
+# Testing Strategy
+
+End-to-end tests, make sure user tap/hover/viewport all works. And the `prefetch` programmatic API works.
+
+# Drawbacks
+
+- Users have limited prefetch features with `@astrojs/prefetch`
+- Users have to use external prefetching solutions
+- Double prefetching could happen as View Transitions prefetches too
+
+# Alternatives
+
+Continue supporting `@astrojs/prefetch`
+
+# Adoption strategy
+
+We should document a migration path for `@astrojs/prefetch` users to use the new `prefetch` option.
+
```

Comment:
Don't we plan to have it as experimental first?

Reply:
I think the plan is to ship it stable immediately. Especially that Astro 4.0 is near if we need to make any change, and that View Transitions that uses this internally is also stable too.